cas-server-webapp/src/main/java/org/jasig/cas/support/pac4j/web/flow/ClientAction.java

```diff
@@ -0,0 +1,254 @@
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected Event doExecute(final RequestContext context) throws Exception {
+        final HttpServletRequest request = WebUtils.getHttpServletRequest(context);
+        final HttpServletResponse response = WebUtils.getHttpServletResponse(context);
+        final HttpSession session = request.getSession();
+
+        // web context
+        final WebContext webContext = new J2EContext(request, response);
+
+        // get client
+        final String clientName = request.getParameter(this.clients.getClientNameParameter());
+        logger.debug("clientName: {}", clientName);
+
+        // it's an authentication
+        if (StringUtils.isNotBlank(clientName)) {
+            // get client
+            final BaseClient<Credentials, CommonProfile> client =
+                    (BaseClient<Credentials, CommonProfile>) this.clients
+                            .findClient(clientName);
+            logger.debug("client: {}", client);
+
+            // Only supported protocols
+            final Mechanism mechanism = client.getMechanism();
+            if (!SUPPORTED_PROTOCOLS.contains(mechanism)) {
+                throw new TechnicalException("Only CAS, OAuth, OpenID and SAML protocols are supported: " + client);
+            }
+
+            // get credentials
+            final Credentials credentials;
+            try {
+                credentials = client.getCredentials(webContext);
+                logger.debug("credentials: {}", credentials);
+            } catch (final RequiresHttpAction e) {
+                logger.debug("requires http action: {}", e);
+                response.flushBuffer();
+                final ExternalContext externalContext = ExternalContextHolder.getExternalContext();
+                externalContext.recordResponseComplete();
+                return new Event(this, "stop");
+            }
+
+            // retrieve parameters from web session
+            final Service service = (Service) session.getAttribute(SERVICE);
+            context.getFlowScope().put(SERVICE, service);
+            logger.debug("retrieve service: {}", service);
+            if (service != null) {
+                request.setAttribute(SERVICE, service.getId());
+            }
+            restoreRequestAttribute(request, session, THEME);
+            restoreRequestAttribute(request, session, LOCALE);
+            restoreRequestAttribute(request, session, METHOD);
+
+            // credentials not null -> try to authenticate
+            if (credentials != null) {
+                final TicketGrantingTicket tgt =
+                    this.centralAuthenticationService.createTicketGrantingTicket(new ClientCredential(credentials));
+                WebUtils.putTicketGrantingTicketInScopes(context, tgt);
+
+                // for institution clients, go to `remoteAuthenticate` and finish authentication through API
+                if (INSTITUTION_CLIENTS.contains(clientName)) {
```

Comment:
This is a straight-forward implementation for OKState. If we have institution authentication through OAuth in the future, this can be re-written in a way which utilize the updated `osf_institution` model.